### PR TITLE
#10127: Move reduce op from tt_lib to ttnn part 1

### PR DESCRIPTION
**Pull request by pavlejosipovic**
Replaced usage of tt_dnn reduced ops in C++ with
ttnn analog.
Replace usage in Python of tt_lib reduce ops with ttnn analog. Remove tt_lib bindings and entries from docs.
Moved cpp code from deprecated/tt_dnn to operations/reduction/generic. Tweaked reduce_scatter to use ttnn.ReduceType enum as input.

Remaining: Code organization in ttnn/cpp/ttnn/operations/reduction/generic/device/

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/demos/t3000/llama2_70b/tt/llama_common.py [hunk 1]
```diff
@@ -212,7 +212,7 @@ def get_weight_cache_path_galaxy(base_cache_path, tensor_str, device_idx, num_de
 def rms_decomp(x, norm_weight, eps):
     squared = ttnn.pow(x, 2)
     # mean_squared = tt_lib.tensor.mean(squared, )
-    sum_squared = tt_lib.tensor.reduce(squared, tt_lib.tensor.ReduceOpMath.SUM, tt_lib.tensor.ReduceOpDim.W, scaler=1.0)
+    sum_squared = ttnn.sum(squared, 3)
```
Comment (ayerofieiev-tt) [inline models/demos/t3000/llama2_70b/tt/llama_common.py:215]: I assume Dim.W is the default one now?

File: models/experimental/bert/fused_ops/layernorm.py [hunk 2]
```diff
@@ -126,13 +123,13 @@ def layernorm_1d_(x, overrideH=None, refx=None, refgamma=None, refbeta=None):
             H_ = overrideH
 
         # first compute the mean (m)
-        means = tensor.reduce(x, RSUM, RW, 1.0 / W)  # -> NCH1
+        means = tensor.sum(x, 3, True, None, None, 1.0 / W)  # -> NCH1
```
Comment (ayerofieiev-tt) [inline models/experimental/bert/fused_ops/layernorm.py:126]: `sum(x, 3, True, None, None, 1.0 / W)`
This call is a bit hard to decode. Unclear what those true/none/none mean w/o knowing the signature of the function. Should those be kw_only args?

File: models/demos/t3000/llama2_70b/tt/llama_common.py (hunk 1 above)
Comment (pavlejosipovic) [reply models/demos/t3000/llama2_70b/tt/llama_common.py:215]: This "3" here represents W dim.

File: models/experimental/bert/fused_ops/layernorm.py [hunk 3]
```diff
@@ -165,17 +162,17 @@ def layernorm_2d_(x):
         W = x.get_legacy_shape()[3]
 
         # first compute the mean (m)
-        redW = tensor.reduce(x, RSUM, RW, 1.0 / W)  # -> NCH1
-        mean = tensor.reduce(redW, RSUM, RH, 1.0)  # -> NC11 (HW reduce doesn't behave well with small scaler)
+        redW = ttnn.sum(x, 3, scalar=1.0 / W)  # -> NCH1
+        mean = ttnn.sum(redW, 2)  # -> NC11 (HW reduce doesn't behave well with small scaler)
```
Comment (ayerofieiev-tt) [inline models/experimental/bert/fused_ops/layernorm.py:166]: These reduced by height previously. In some cases we removed Dim.W, here we basically remove Dim.H. 
Is it intentional?

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp [hunk 4]
```diff
@@ -29,128 +26,11 @@ template <ReduceType reduce_type>
 struct Reduce {
     static Tensor operator()(
         const Tensor& input_tensor_arg,
-        const std::optional<std::variant<int, std::vector<int>>>& dim_arg,
-        const bool keepdim,
+        const std::optional<std::variant<int, std::vector<int>>>& dim_arg = std::nullopt,
+        const bool = true,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp:30]: please review this

**Review by ayerofieiev-tt — APPROVED** [review]
Overall looks good

File: models/demos/t3000/llama2_70b/tt/llama_common.py (hunk 1 above)
Comment (ayerofieiev-tt) [reply models/demos/t3000/llama2_70b/tt/llama_common.py:215]: Thank you

File: models/experimental/bert/fused_ops/layernorm.py (hunk 3 above)
Comment (pavlejosipovic) [reply models/experimental/bert/fused_ops/layernorm.py:166]: This should be functionally equivalent, first line was reduce RW which now maps to 3 (dim) and the sencond line was RH which now maps to "2"

File: models/experimental/bert/fused_ops/layernorm.py (hunk 2 above)
Comment (pavlejosipovic) [reply models/experimental/bert/fused_ops/layernorm.py:126]: yep, I'll change this one

File: ttnn/cpp/ttnn/operations/reduction/reduction_pybind.hpp [hunk 5]
```diff
@@ -19,6 +19,8 @@
 namespace ttnn::operations::reduction {
 
 void py_module(py::module& module) {
+    export_enum<ttnn::operations::reduction::ReduceType>(module, "ReduceType");
```
[the remark this answers is not in the document]
Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/reduction/reduction_pybind.hpp:22]: sure, will do

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp (hunk 4 above)
Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp:30]: thx, I'm not sure how I ended up with this

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions_pybind.hpp [hunk 6]
```diff
@@ -27,7 +27,8 @@ void bind_reduction_operation(py::module& module, const reduction_operation_t& o
             py::arg("dim") = std::nullopt,
             py::arg("keepdim") = true,
             py::arg("memory_config") = std::nullopt,
-            py::arg("compute_kernel_config") = std::nullopt});
+            py::arg("compute_kernel_config") = std::nullopt,
+            py::arg("scalar") = 1.0f});
```
[the remark this answers is not in the document]
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions_pybind.hpp:31]: for everything past keepdim make it keyword only on the python side by doing this:

ttnn::pybind_arguments_t{
            py::arg("input_tensor"),
            py::arg("dim") = std::nullopt,
            py::arg("keepdim") = true,
            **py::kw_only(),**
            py::arg("memory_config") = std::nullopt,
            py::arg("compute_kernel_config") = std::nullopt});
            py::arg("compute_kernel_config") = std::nullopt,
            py::arg("scalar") = 1.0f});

**Review by sjameelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp (hunk 4 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.hpp]: happens

**Review by djordje-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions_pybind.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions_pybind.hpp]: thx

**Review by cglagovichTT — APPROVED** [review]

File: tests/tt_eager/ops/test_reduce_op.cpp [hunk 7]
```diff
@@ -1,114 +0,0 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
-//
-// SPDX-License-Identifier: Apache-2.0
```
Comment (tt-aho) [inline tests/tt_eager/ops/test_reduce_op.cpp:3]: Are we just deleting this test, or was it moved elsewhere?

Comment (pavlejosipovic) [reply tests/tt_eager/ops/test_reduce_op.cpp:3]: Just delete, along with the reduce op from tt_lib

**Comment by pavlejosipovic** [comment]
T3k frequent test:  https://github.com/tenstorrent/tt-metal/actions/runs/10367119174
All post commit: https://github.com/tenstorrent/tt-metal/actions/runs/10367076268
Model perf regression: https://github.com/tenstorrent/tt-metal/actions/runs/10367076268
Device perf regressions and output report: https://github.com/tenstorrent/tt-metal/actions/runs/10367099635
All of the above are green.

t3k demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/10367105501
Falcon 40B is failing but same failure is happening on main.

File: tests/tt_eager/ops/test_reduce_op.cpp (hunk 7 above)
Comment (tt-aho) [reply tests/tt_eager/ops/test_reduce_op.cpp:3]: Shouldn't we re-add the same test for the ttnn version to keep existing coverage? Or is this a redundant test?

Comment (pavlejosipovic) [reply tests/tt_eager/ops/test_reduce_op.cpp:3]: We have tests in ttnn, these tests were not great, they would just run things w/o checking if results are valid in any way. Check the comment on line 30

**Pull request merged**